### ISSUE TITLE
Support for Microsoft Edge?

### ISSUE THREAD

**Issue by rogersachan** [ipfs-shipyard/ipfs-companion#353]

**Comment by lidel** [comment]
Good question :)  There is an open task in #276 for smoke-testing it under MS Edge, but nobody picked it up yet (most of our community seems to be using Firefox/Chrome).

@rogersachan  do you have MS Edge and some spare time on your hands?  If so, try loading [ipfs_companion-2.1.0.zip](https://github.com/ipfs-shipyard/ipfs-companion/releases/download/v2.1.0/ipfs_companion-2.1.0.zip) (see [this instruction](https://docs.microsoft.com/en-us/microsoft-edge/extensions/guides/adding-and-removing-extensions))  and let us know if there were any errors (AFAIK Edge does not support notifications API yet, so you may need to inspect Browser Console). 

It would be an important data point telling us where MS Edge stands.

**Comment by rogersachan** [comment]
![image](https://user-images.githubusercontent.com/7173984/34651264-ff2cd90e-f39b-11e7-8e72-817cf7993047.png)
I get an error before I can even load this extension. There are no other errors or messages in the console.

**Comment by rogersachan** [comment]
![image](https://user-images.githubusercontent.com/7173984/34651319-a5a727c6-f39c-11e7-8202-89d7c247134a.png)
Updating background with this in order to fix the issue gives me another error.
![image](https://user-images.githubusercontent.com/7173984/34651326-c41a6736-f39c-11e7-9087-7b4d58825b9d.png)

**Comment by rogersachan** [comment]
Deleting both 'browser_style' fields allowed me to import the extension. I'll start testing it as soon as I can.

**Comment by rogersachan** [comment]
Attempting any of the links in #312 just sends me to the search engine.

**Comment by rogersachan** [comment]
The background page is now giving me this error: 
```
Unable to initialize addon due to error Error: Invocation of form storage.local.get(object) doesn't match definition storage.local.get(optional string or array or object keys, function callback).
```

**Comment by lidel** [comment]
This is a _very_ good feedback that cleared a lot! 👍 🔍 
Thank you @rogersachan!

Different handling of manifest keys is inconvenient, but can be managed. 
But the last error is a show-stopper.

It seems that Edge provides `browser.*` APIs as callback-only. On top of that presence of `browser` object  prevents  [webextension-polyfil](https://github.com/mozilla/webextension-polyfill) from being correctly applied.  This is very unfortunate, they should either expose callback-only APIs under `chrome.*` namespace, or add promise support to `browser.*`. 

There is an interesting discussion at https://github.com/mozilla/webextension-polyfill/issues/3.
The [upstream ticket at Microsoft](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9421085/) about adding promise support was just reopened. 

I am marking it as "blocked" until it is resolved upstream, either by [Microsoft](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9421085/) or [Mozilla](https://github.com/mozilla/webextension-polyfill/issues/3).

**tl;dr**
 > At the moment, neither a working polyfill, nor any other cross browser solution exists. 
> – https://github.com/mozilla/webextension-polyfill/issues/3#issuecomment-335331528

>  We hear the request for this support. Have reactivated this request, and added to our backlog.
>  – [Microsoft Edge Team](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9421085/#comment-7)

**Comment by cvan** [comment]
@lidel: @rogersachan: Good catches here.

Perhaps I'm missing some context in mozilla/webextension-polyfill#3, but couldn't `browser.storage` be special-cased and handled in the `webextension-polyfill` [à la the [`browser.storage.onRuntime`](https://github.com/mozilla/webextension-polyfill/blob/47ddfbf/src/browser-polyfill.js#L329-L349) patch)?

Or is it more complicated than that?

**Comment by lidel** [comment]
@cvan I am afraid fixing `browser.storage` alone is not enough: another error caused by the lack of promise support will pop up elsewhere. 

I did not look at internals of `webextension-polyfill`, but my intuition is that that polyfills for entire `browser.*` would have to be sprinkled with code responsible for Edge-case (accidental pun) detection.   

Unless it can be solved in one swoop globally, not sure if it is worth the effort, especially now that [upstream ticket got reopened](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9421085/).

**Comment by cvan** [comment]
> I did not look at internals of `webextension-polyfill`, but my intuition is that that polyfills for entire `browser.*` would have to be sprinkled with code responsible for Edge-case (accidental pun) detection.

I just had to comment that I laughed out loud at this. I will look into this (and the relevant [Edge ticket](https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9421085/)).

**Comment by lidel** [comment]
Just dropping a quick note that Edge is going to be likely supported once it has been switched to Chromium backend. Let's keep this issue open until that happens.

Refs.
- https://blog.mozilla.org/blog/2018/12/06/goodbye-edge/
- https://github.com/mozilla/webextension-polyfill/pull/114#issuecomment-451923835

**Comment by rogersachan** [comment]
The new Microsoft Edge which is built on chromium (https://www.microsoftedgeinsider.com/en-us/) now supports chrome extensions from the chrome web store, this issue is no longer needed.

**Comment by lidel** [comment]
@rogersachan thank you for letting us know! 
@hacdias are you able to confirm extension loads as expected?

**Comment by hacdias** [comment]
@lidel, seems to be running as expected :smile:

![Screenshot 2019-04-12 at 23 15 33](https://user-images.githubusercontent.com/5447088/56069007-e91d9700-5d78-11e9-92d3-628000412d04.png)
